### PR TITLE
Wh fix

### PR DESCRIPTION
**Pull request by vydd**

### CODE REVIEW

File: src/sketch.lisp [hunk 1]
```diff
@@ -182,136 +182,126 @@ used for drawing, 60fps.")
...
+							  ,(binding-value b)))
+			when (funcall (binding-in custom-bindings) b)
+			  collect `(,(binding-name b) (or (getf initargs ,(binding-initarg b))
+							  ,(binding-value b)))))
+	   (declare (ignorable ,@(mapcar #'binding-name all-bindings)))
+	   (setf ,@(loop for b in all-bindings
+			 collect `(,(binding-accessor b) *sketch*)
+			 collect (binding-name b))))
+	 (setf (env-y-axis-sgn (slot-value *sketch* '%env))
+	       (if (eq (slot-value *sketch* 'y-axis) :down) +1 -1))))))
```
Comment (Gleefre) [inline src/sketch.lisp:293]: This `setf` actually is not needed, as it should be handled by `(setf (sketch-x-axis ...) ...)`

File: src/sketch.lisp [hunk 2]
```diff
@@ -182,136 +182,126 @@ used for drawing, 60fps.")
...
+		; We need to use the order from defined-bindings to support let*-like behavior
+		,@(loop for b in defined-bindings
+			when (funcall (binding-in changed-bindings) b)
+			  collect `(,(binding-name b) (if (getf initargs ,(binding-initarg b))
+							  (slot-value *sketch* ',(binding-name b))
+							  ,(binding-value b)))
+			when (funcall (binding-in custom-bindings) b)
+			  collect `(,(binding-name b) (or (getf initargs ,(binding-initarg b))
+							  ,(binding-value b)))))
+	   (declare (ignorable ,@(mapcar #'binding-name all-bindings)))
```
Comment (Gleefre) [inline src/sketch.lisp:288]: Hm, actually this `declare` is uncalled for - all variables are used in the `setf` below.

File: src/sketch.lisp [hunk 3]
```diff
@@ -182,136 +182,126 @@ used for drawing, 60fps.")
...
+	  (unchanged-bindings (remove-if (binding-in defined-bindings) default-bindings)))
+      `(defmethod prepare ((*sketch* ,sketch-name) &rest initargs &key &allow-other-keys)
+	 (declare (ignorable initargs))
+	 (let* (,@(loop for b in unchanged-bindings
+			collect `(,(binding-name b) (slot-value *sketch* ',(binding-name b))))
+		; We need to use the order from defined-bindings to support let*-like behavior
+		,@(loop for b in defined-bindings
+			when (funcall (binding-in changed-bindings) b)
+			  collect `(,(binding-name b) (if (getf initargs ,(binding-initarg b))
+							  (slot-value *sketch* ',(binding-name b))
```
Comment (Gleefre) [inline src/sketch.lisp:283]: Actually, there is no need to make a distinction here: `initargs` will be non null only when called from `initialize-instance`. In this case the value of `(getf initargs ,(binding-initarg b))` will be the same as the `(slot-value *sketch* ',(binding-name b))` as it the slot was just set to that value by the executed `initialize-instance`.

File: src/sketch.lisp [hunk 4]
```diff
@@ -182,136 +182,126 @@ used for drawing, 60fps.")
...
+
+(defun define-prepare-method (sketch-name default-bindings defined-bindings custom-bindings all-bindings)
+  (flet ((binding-in (bindings)
+	   (lambda (binding) (member binding bindings :test #'binding=))))
+    (let ((changed-bindings (remove-if-not (binding-in default-bindings) defined-bindings))
+	  (unchanged-bindings (remove-if (binding-in defined-bindings) default-bindings)))
+      `(defmethod prepare ((*sketch* ,sketch-name) &rest initargs &key &allow-other-keys)
+	 (declare (ignorable initargs))
+	 (let* (,@(loop for b in unchanged-bindings
+			collect `(,(binding-name b) (slot-value *sketch* ',(binding-name b))))
```
Comment (Gleefre) [inline src/sketch.lisp:278]: For `unchanged-bindings` you could use the same "formula" as for `changed-bindings` with the default value `(binding-value b)` being a literal - string, symbol or a number - so it won't be executed twice for no reason.

**Review by Gleefre — COMMENTED** [review]
About `prepare`: in general, using `&rest initargs` parameter to parse keyword arguments is not efficient - and it happens at runtime too. It would be better to leave such parsing to the lisp implementation.

Note that all bindings can be simplified to
```lisp
`(or (getf initargs ,(binding-initarg b))
     ,(binding-value b))
```
This is exactly what the default value for keyword arguments does, so all bindings could be simplified even further if written as `&key` arguments to the `prepare` method.

**Comment by vydd** [comment]
Solved in #87

**Pull request closed without merging**